protocol.md

```diff
@@ -0,0 +1,249 @@
+
+Whisper adapter specifies interaction with the Whiper service with regards to keys management, configuration and some metadata (like topic) required to properly process and encrypt/decrypt messages.
+
+Offline messaging describes how the protocol handles delivering messages when one or more participants were offline and the messages expired in the network.
+
+The protocol does not specify things like peers discovery, however, some notes and the current implementation will be described in an appendix.
+
+# Payload
+
+Payload is encrypted and decrypted using [transit format](https://github.com/cognitect/transit-format). It is a text-based format so that each decrypted message's payload can be easily read.
+
+Payload contains the following fields:
+* text `string`
+* content type `enum { text/plain }`
+* message type `enum { public-group-user-message, user-message }`
+* clock `int64`
+* timestamp `int64`
+* content `struct { chat-id string, text string }`
+
+Example of a valid message:
+
+```
+["~#c4",["abc123","text/plain","~:public-group-user-message",154593077368201,1545930773682,["^ ","~:chat-id","testing-adamb","~:text","abc123"]]]
+```
+
+As you can see, the message is an array and each index value has its meaning:
+* 0: `c4` is a decoder handler identificator for the current payload format. Identificators allow to register handler for many different types of messages
+* 1: array which items correspond to the described payload fields above
+
+For more details regarding serialization and deserialization please consult [transit format](https://github.com/cognitect/transit-format) specification.
+
+## Content types
+
+Here is a list of supported content types:
+* `text/plain` identifies a message which content is a plain text
+
+TODO: are there other types? How is it useful? Sending Eth requests?
+
+## Message types
+
+Here is a list of supported message types:
+* `public-group-user-message` is a message to the public group
+* `user-message` is a private message.
+
+TODO: how is it useful? What are other supported message types?
+
+## Clock vs Timestamp And Message Ordering
+
+`timestamp` is Unix time calculated when the message is created (TODO: does it come from Whisper or the system?)
+
+`clock` is calculated using the algorithm of [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamps). When there are messages available in a chat, `clock`'s value is calculated as `last-message-clock-value + 1`. If there are no messages, `clock` is initialized with `timestamp`'s value.
+
+`clock` value is used for the message ordering. Due to the used algorithm, distributed nature of the system and physics, we achieve casual ordering which might produce counterintuitive results in some edge cases. For example, when one joins a public chat and sends a message before receiving the exist messages, their message `clock` value might be lower and the message will end up in the past when the historical messages are fetched.
+
+## Quoting
+
+TBD
```

Comment:
@yenda @rasom and this one.

Reply:
I don't understand what is suppose to be put there? is it about the replies?

Reply:
Think so

Reply:
Yes, it's how replies are handled.